@here/harp.gl/lib/index.ts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export * from "@here/harp-geoutils/lib/math/Box3Like";
+
+export {
+    MapView,
+    DataSource,
+    Tile,
+    MapViewEventNames,
+    MapViewOptions,
+    CopyrightElementHandler
+} from "@here/harp-mapview";
+export * from "@here/harp-omv-datasource";
+export * from "@here/harp-geojson-datasource";
+export * from "@here/harp-webtile-datasource";
```

Comment:
This looks quite arbitrary :/ Can we split it a bit more? E.g. have the datasources as plugins that just add to the global `harp` namespace?

Kinda like momentjs.com does with timezones and holidays, e.g. https://momentjs.com/timezone/ and https://www.npmjs.com/package/moment-holiday

Reply:
This is how our API looks like _now_.
Moreover, we have _no_ chance to create pluggable system without turning into SystemJS which is complex and/or modules which are not yet ready for wide audienc.

> have the datasources as plugins 

Which? ATM we have actually one datasource i.e omv-datasource which is our _de_facto_ only one real, supported and featured datasource.
* `WebTileDatasource` is only technology preview.
* `GeoJson...` is being deprecated as OmvDataSource will support geojson.

Part of this story is "make it simple". And simple means 1 or max 2 scripts (including 3JD)
If we will end with
```
<script src="https://unpkg.com/@here/harp.gl-core"></script>
<script src="https://unpkg.com/@here/harp.gl-core"></script>

`moment-holiday.js` works because `moment` by default installs itself in global namespace.
We can have `harp` as root - but what will it contain ?
`MapView` - ok

* `harp.controls` as alias for stuff in `@here/harp-map-controls`
* `harp.omvDatasource` or harp.OmvDatasource` and thus `harp.OmvDatasource.OmvDataSource` ?

Our current module system is and naming is so UMD non-friendly, that it cannot be fixed in scope of such small task. Actually, it's not modular at all.

Reply:
This is how our API looks like _now_.
Moreover, we have _no_ chance to create pluggable system without turning into SystemJS which is complex and/or modules which are not yet ready for wide audienc.

> have the datasources as plugins 

Which? ATM we have actually one datasource i.e `omv-datasource` which is our _de_facto_ only one real, supported and featured datasource.
* `WebTileDatasource` is only technology preview.
* `GeoJson...` is being deprecated as OmvDataSource will support geojson.

Part of this story is "make it simple". And simple means 1 or max 2 scripts (including 3JS)
If we will end with
```
<script src="https://unpkg.com/@here/harp.gl-core"></script>
<script src="https://unpkg.com/@here/harp.gl-omv-datawhatever"></script>
<script src="https://unpkg.com/@here/harp.gl-map-controls"></script>
```
this is far from being simple.

Going further, 
`moment-holiday.js` works because `moment` is well designed, Javascript first API which is designed to be UMD/IIFE compatible.
It  installs itself in global namespace since always and by default.

Our current module system is and naming is so UMD non-friendly, that it cannot be fixed in scope of such small task. Actually, it's not modular at all.

Reply:
I'd argue that all 'layers' are plugins:

```
<script src="https://unpkg.com/@here/harp.gl-core"></script>
<script src="https://unpkg.com/@here/harp.gl-omv-datawhatever"></script>
```

This would mean two scripts, and still give us the flexibility to add other layers on top.

Reply:
> UMD ...

It's all about UMD or not UMD.
Some time ago we dropped  UMD as not supported.
UMD works if we're can import everything efficiently from each-other using `index.ts` or `index-decoder.ts`  etc.

That means, no imports like this:
```
import { LineGroup } from "@here/harp-lines/lib/Lines";
```
Only this:
```
import { LineGroup } from "@here/harp-lines";
```
But then, `decoder` will get be bundled with whole `lines` module including it's deps like materials and TREE.js.

First thing that we have to do is a proper module division, that is built around key concept - don't leak code between UI thread and decoders.

Reply:
> harp.gl-omv-datawhatever

Actually, i'd like to name it `harp.gl-omv-tile-provider`.

Problem with this particular is that `OmvDataSource` and friends which are loaded in main thread is like "5 kilobytes" of pure wrapper over `TileDataSource` (`OmvDataSource is basically `TileDataSource`) and basic `fetch`-based `OmvRestSomethingProvider`.

Same applies to `WebTileDataSource` - 1-2k minified, no more.
So datasources don't weight much in `main bundle` of `harp.js`

The main part that weights is - decoders and specifically, `OmvDecoder` and in future, TerrainDecoder whose loads tons of 3rdparty-libs. 
Ideally, `decoders` script should be split also so user has another bundle/script for all the services in workers:

 * harp-omv-decoder-worker.js for OmvDecoderService
 * harp-tiler-worker.js for TilerService
 * harp-terrain-decoder-worker.js for TerrainDecoderService

Now, each of them requires (usually only math! :/) part of 3JS and tons of common code, so we should avoid creating 4 * eachService workers, because we would end with at least 12 workers in common case (4 for omv, 4 for tiler, 4 for terrain).

So, we wan't to load all the services in each worker (to reuse JS runtimes and memory), ...
... worker scripts should be "UMD/AMD"-like so we can load them using `requirejs`/systemd like semantics.
That leaves us with following "architecture":
 * `three.js`
 * `harp-worker-core.js` (depends on 3js probably)
 * `harp-omv-decoder-worker.js` (depends on 3js, proto,  harp-worker-core.js)
 * `harp-tiler-worker.js` (depends on 3js,  harp-worker-core.js and some custom stuff)
 * `harp-terrain-decoder-worker.js` (depends on 3js, proto,  harp-worker-core.js)

That whole issue is can of worms literally ...